fix some invalidations when loading Static.jl

ranocha opened the pull request:
This improvement prevents some invalidations in abstractarray.jl when loading Static.jl.

This is based on the following code:

```julia
julia> using Pkg; Pkg.activate(temp=true); Pkg.add("Static")

julia> using SnoopCompileCore; invalidations = @snoopr(using Static); using SnoopCompile

julia> trees = invalidation_trees(invalidations)
inserting !(::False) in Static at ~/.julia/packages/Static/sVI3g/src/Static.jl:427 invalidated:
...
                 51: signature Tuple{typeof(!), Any} triggered MethodInstance for Base.mightalias(::Vector, ::Vector{Any}) (10 children)
                 52: signature Tuple{typeof(!), Any} triggered MethodInstance for Base.mightalias(::Vector, ::Vector{String}) (11 children)
                 57: signature Tuple{typeof(!), Any} triggered MethodInstance for Base.mightalias(::Vector, ::Vector{Pkg.Types.PackageSpec}) (16 children)
...
```

I suggest the labels `latency` and `backport-1.8`.

ranocha committed:
improve type stability of `mightalias(A::AbstractArray, B::AbstractArray)`

This improvement prevents some invalidations in abstractarray.jl when loading Static.jl.

vtjnash reviewed (changes requested):
The compiler probably should be able to figure this out

ranocha commented:
But it looks like it does not at the moment, at least not on Julia v1.8. Is it already fixed on `master`? If not, I think it would be nice to get this merged so that many people can benefit from reduced latency.

timholy commented:
I would try to fix this at a higher level; presumably this fails to infer only because the array type is being inferred abstractly.

ranocha commented:
> I would try to fix this at a higher level; presumably this fails to infer only because the array type is being inferred abstractly.

Yes, the first argument of `mightalias` is only inferred as `Vector` in these cases. I will try to see whether I can fix it at a higher level.

ranocha committed:
Revert "improve type stability of `mightalias(A::AbstractArray, B::AbstractArray)`"

This reverts commit 6b3a3ae733d71364a44c948b1a655db3a153bfb5.

ranocha committed:
improve type stability of `findall(testf::Function, A::AbstractArray)`

This prevents some invalidations in `mightalias(A::AbstractArray, B::AbstractArray)` in abstractarray.jl when loading Static.jl.

ranocha commented:
The last two cases reported above come from `findall(testf::Function, A::AbstractArray)`, which used broadcasting. I changed it to `map`, removing most invalidations.

The first case (with 10 children) is reduced to 7 children:
```
                 134: signature Tuple{typeof(!), Any} triggered MethodInstance for Base.mightalias(::Vector, ::Vector{Any}) (7 children)
```
I could not really find a good place where the type instability happens there since `ascend` ends already with `copyto!(::Vector, ...)`:
```
julia> ascend(root)
Choose a call for analysis (q to quit):
     mightalias(::Vector, ::Vector{Any})
       unalias(::Vector, ::Vector{Any})
         broadcast_unalias(::Vector, ::Vector{Any})
           preprocess(::Vector, ::Vector{Any})
             preprocess_args(::Vector, ::Tuple{Vector{Any}})
               preprocess_args(::Vector, ::Tuple{Base.RefValue{String}, Vector{Any}})
                 preprocess(::Vector, ::Base.Broadcast.Broadcasted{Nothing, Tuple{Base.OneTo{Int64}}, Pkg.API.var
 >                 copyto!(::Vector, ::Base.Broadcast.Broadcasted{Nothing, Tuple{Base.OneTo{Int64}}, Pkg.API.var"
```

N5N3 commented on base/array.jl:
```diff
@@ -2325,7 +2325,7 @@ findall(testf::Function, A) = collect(first(p) for p in pairs(A) if testf(last(p
 # Broadcasting is much faster for small testf, and computing
 # integer indices from logical index using findall has a negligible cost
-findall(testf::Function, A::AbstractArray) = findall(testf.(A))
+findall(testf::Function, A::AbstractArray) = findall(map(testf, A))
```
This would increase the allocation size as map return a `Array{Bool}` by default while `broadcast` return a `BitArray`.

ranocha replied:
Okay, didn't know that. Thanks for noticing!

ranocha committed:
improve type stability of `findall(testf::Function, A::AbstractArray)`

This prevents some invalidations in `mightalias(A::AbstractArray, B::AbstractArray)` in abstractarray.jl when loading Static.jl.
Here we specialize on the function instead of using map since broadcasting returns a BitArray, map a Vector{Bool}.

ranocha commented on base/array.jl:
```diff
@@ -2325,7 +2325,7 @@ findall(testf::Function, A) = collect(first(p) for p in pairs(A) if testf(last(p
 # Broadcasting is much faster for small testf, and computing
 # integer indices from logical index using findall has a negligible cost
-findall(testf::Function, A::AbstractArray) = findall(testf.(A))
+findall(testf::F, A::AbstractArray) where {F<:Function} = findall(testf.(A))
```
This is another possibility to fix the invalidations coming from this by specializing on `testf`.

vtjnash commented:
@aviatesk Could we get Conditional to refine this `Type{<:AbstractArray}` to `DataType` because of the explicit `isa` test? I suspect that would generally reflect better what the user is intending to do with it (though it will mean the tmerge at the phi after the if branch will widen again to just Type)

timholy commented:
I suspect this could deserve a long-ish conversation. I've often wondered about making this change (mostly in `map` but here in `filter` too), and wondered what the tradeoff might be. Not having the function be inferrable causes lots of headaches, as you've discovered here. But as you say, specialization contributes to latency.

Previously, my inclination was to wait until precompilation got better, because good precompilation circumvents the latency. But we may have basically arrived there, or will soon once we can cache native code. So maybe the time is ripe for this conversation. There are other issues where we might consider changing our verdict, e.g., #36454.

So I guess the key question is, are there reasonable benchmarks we could use to help make this decision?

ranocha commented:
> So I guess the key question is, are there reasonable benchmarks we could use to help make this decision?

I don't think I have really good benchmarks for this. I could of course provide some biased benchmarks from our packages such as Trixi.jl, but we will need more diverse input, I think. In the end, there is no single best benchmark for everybody but we should look at something reasonable for most people.
